File: day7_part2_2016.js
const fs = require('fs');
const readline = require('readline');
const regex = /\[[a-z]+\]/g;

const readInterface = readline.createInterface({
    input: fs.createReadStream('input.txt'),
    output: process.stdout,
    console: false
});

let sslCount = 0;

readInterface.on('line', function(line) {
    if (supportsSSL(line)) {
        sslCount++;
    }
});

readInterface.on('close', function() {
    console.log(sslCount);
});

function supportsSSL(ip) {
    const bracketContents = ip.match(regex) || [];

    ip = ip.replace(regex, "-");
    const abas = findABAs(ip);

    for (const aba of abas) {
        const bab = aba[1] + aba[0] + aba[1];
        for (const bracketContent of bracketContents) {
            if (bracketContent.includes(bab)) {
                return true;
            }
        }
    }

    return false;
}

function findABAs(s) {
    const abas = [];
    for (let i = 0; i < s.length - 2; i++) {
        if (s[i] !== s[i + 1] && s[i] === s[i + 2]) {
            abas.push(s.substring(i, i + 3));
        }
    }
    return abas;
}

File: go/day7_part2_2016.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	sslCount := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if supportsSSL(scanner.Text()) {
			sslCount++
		}
	}

	fmt.Println(sslCount)
}

func supportsSSL(ip string) bool {
	insideBrackets := regexp.MustCompile(`\[[a-z]+\]`)
	bracketContents := insideBrackets.FindAllString(ip, -1)

	// Generate BABs from ABAs found outside brackets
	ip = insideBrackets.ReplaceAllString(ip, "-")
	for _, aba := range findABAs(ip) {
		bab := string([]rune{rune(aba[1]), rune(aba[0]), rune(aba[1])})
		for _, bracketContent := range bracketContents {
			if strings.Contains(bracketContent, bab) {
				return true
			}
		}
	}

	return false
}

func findABAs(s string) []string {
	var abas []string
	for i := 0; i < len(s)-2; i++ {
		if s[i] != s[i+1] && s[i] == s[i+2] {
			abas = append(abas, s[i:i+3])
		}
	}
	return abas
}